[13.0] stock_storage_type: Optimize "will contain" computed fields on stock.location

guewen opened the pull request:
The 2 fields:

* location_will_contain_product_ids
* location_will_contain_lot_ids

Are used only when applying a related location storage type with
"do_not_mix_products" or "do_not_mix_lots". As the result of the
computed fields are stored in m2m related fields, we can prevent a
lot of insertions in these relation tables by not computing them
if they have no location storage type that will need them.

ref: 1964

Commit:
Optimize "will contain" computed fields on stock.location

The 2 fields:

* location_will_contain_product_ids
* location_will_contain_lot_ids

Are used only when applying a related location storage type with
"do_not_mix_products" or "do_not_mix_lots". As the result of the
computed fields are stored in m2m related fields, we can prevent a
lot of insertions in these relation tables by not computing them
if they have no location storage type that will need them.

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -141,21 +141,48 @@ def _compute_leaf_location_ids(self):
-    @api.depends("quant_ids", "in_move_ids", "in_move_line_ids")
+    def _should_compute_will_contain_product_ids(self):
+        return any(
+            location.do_not_mix_products
```
No need to compute if "do not mix lot" is used as the product cache table won't be used but the lot cache table
```suggestion
            location.do_not_mix_products and not location.do_not_mix_lots

```

guewen replied:
I'll change the domain (https://github.com/OCA/wms/blob/5a44b7c04fcbe283ae0b36dbc3653769f127fb6b/stock_storage_type/models/stock_location_storage_type.py#L138-L156) the same way otherwise it is not obvious to understand why the field is empty but still in the domain.

guewen replied:
@jbaudoux after the change, tests do not pass, and I think it's normal: even if a location do not mix lots, some products may not be using lots, so we still need to cache the products and check for both.

guewen replied:
Then we could store only products tracked by lot in one case and those not in the other... but I don't think it's worth it re. added complexity vs the gain.

jbaudoux replied:
@guewen Good point

simahawk commented on stock_storage_type/models/stock_location.py:
```diff
@@ -141,21 +141,48 @@ def _compute_leaf_location_ids(self):
-                | rec.mapped("in_move_ids.product_id")
-                | rec.mapped("in_move_line_ids.product_id")
-            )
+            products = self.env["product.product"].browse()
```
do you expect to have many records here? If yes, `set` of ids would be better... as you know :wink: 
Same below.

guewen replied:
We have 3 sets of products to union (quants, incoming moves, incoming move lines), what's slow is iterating thousands of times to union record id(s). (`browse()` will be called only 2 times: https://github.com/odoo/odoo/blob/36d5bc3c401b2c5f569410108c0a6c8be63818a9/odoo/models.py#L5659-L5664)

jbaudoux reviewed (commented):
Is there a real benefit of this cache?
Is it recomputed, on a location, each time a quant qty is modified, so at each picking (out) move? So probably more often than we put something on it.
Is it recomputed, each time a put-away has selected this location (so we have created a in move line to that location)? So we gain in the computation of the put-away but we still have to refresh the cache afterwards?

There is also something conceptually wrong I think in the approach of what we try to achieve and that could question the need of the cache.
I see 2 strategies for put-away:
* empty bin: we select a new location
* do not mix: we should here select a location that already contains the product (the one with the less qty) or a new one if none is found. Currently, I have the impression that it is building a set that contains empty locations + product's locations and taking one randomly.

guewen commented:
You are probably right that it is not really an overall perf improvement. Maybe removing the store is enough. We should test.

However for this:  



>     * do not mix: we should here select a location that already contains the product (the one with the less qty) or a new one if none is found. Currently, I have the impression that it is building a set that contains empty locations + product's locations and taking one randomly.

I must say I'm baffled: this is the very first time I hear or read your definition in any specification of this feature (and by reading again, I see no trace). So yes, currently, it follows the same rule if this is the first time you putaway a product or the second time. 
Also, unless we are sure we have all the constraints in place (number of items, height, weight, ...), if we always select the location where we have the less product, we risk to frequently put them in over-capacity.

jgrandguillaume commented:
@guewen  This is because @jbaudoux speaks too fast ;) Joke apart, I think do not mix product is before anything else responsible to "not mix product" ! So we want to be sure the location we found is containing the product or we take a new one. Taking the min qty between location containing a same product would be nice to have.

guewen commented:
FTR, when I run the "will_contain" tests with:

`py.test --odoo-config=odoorc --odoo-database=storage_buffer_test -s --disable-warnings wms/stock_storage_type/tests -k will_contain --odoo-log-level=debug_sql`

When the fields `location_will_contain_product_ids`, and `location_will_contain_lot_ids` are stored, we get:

```
.2021-01-22 14:31:40,335 30530 DEBUG storage_buffer_test odoo.sql_db: SQL LOG from:
2021-01-22 14:31:40,335 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_location_stock_production_lot_rel: 0:00:00.000195/1
2021-01-22 14:31:40,335 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_quant_package: 0:00:00.000436/1
2021-01-22 14:31:40,335 30530 DEBUG storage_buffer_test odoo.sql_db: table: product_packaging: 0:00:00.000389/2
2021-01-22 14:31:40,335 30530 DEBUG storage_buffer_test odoo.sql_db: table: product_product_stock_location_rel: 0:00:00.000524/2
2021-01-22 14:31:40,335 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_picking: 0:00:00.000700/2
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_package_storage_type: 0:00:00.000700/2
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: res_users: 0:00:00.001381/2
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: res_partner: 0:00:00.002038/2
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_inventory_line: 0:00:00.001301/3
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: res_lang: 0:00:00.001319/3
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_warehouse: 0:00:00.001410/3
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: ir_config_parameter: 0:00:00.000938/4
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_location_route: 0:00:00.001141/4
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: product_category: 0:00:00.001176/4
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_location_storage_type: 0:00:00.001051/5
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_storage_location_sequence: 0:00:00.001185/5
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_move: 0:00:00.001737/5
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: product_template: 0:00:00.001908/5
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: ir_sequence: 0:00:00.002425/6
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: product_product: 0:00:00.002245/7
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_move_line: 0:00:00.003570/8
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_rule: 0:00:00.001860/9
2021-01-22 14:31:40,336 30530 DEBUG storage_buffer_test odoo.sql_db: table: ir_model_fields: 0:00:00.002148/9
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: uom_uom: 0:00:00.002184/9
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: decimal_precision: 0:00:00.001848/10
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_quant: 0:00:00.003283/10
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_picking_type: 0:00:00.002887/11
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: ir_property: 0:00:00.005348/12
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_location: 0:00:00.015007/60
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: ir_model_data: 0:00:00.013566/64
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: SUM from:0:00:00.075900/410 [422]
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: SQL LOG into:
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_location_allowed_location_storage_type_rel: 0:00:00.000396/1
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_location: 0:00:00.000789/1
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_rule: 0:00:00.001965/1
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_location_stock_production_lot_rel: 0:00:00.000707/2
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: product_packaging: 0:00:00.001086/2
2021-01-22 14:31:40,337 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_quant: 0:00:00.001338/2
2021-01-22 14:31:40,338 30530 DEBUG storage_buffer_test odoo.sql_db: table: ir_property: 0:00:00.001438/2
2021-01-22 14:31:40,338 30530 DEBUG storage_buffer_test odoo.sql_db: table: product_product: 0:00:00.001568/2
2021-01-22 14:31:40,338 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_move_line: 0:00:00.002272/2
2021-01-22 14:31:40,338 30530 DEBUG storage_buffer_test odoo.sql_db: table: product_template: 0:00:00.002859/2
2021-01-22 14:31:40,338 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_production_lot: 0:00:00.003052/2
2021-01-22 14:31:40,338 30530 DEBUG storage_buffer_test odoo.sql_db: table: stock_move: 0:00:00.004135/3
2021-01-22 14:31:40,338 30530 DEBUG storage_buffer_test odoo.sql_db: table: product_product_stock_location_rel: 0:00:00.001385/4
2021-01-22 14:31:40,338 30530 DEBUG storage_buffer_test odoo.sql_db: SUM into:0:00:00.022989/410 [422]
```

And when we do not store them:

```
2021-01-22 14:36:18,182 31973 DEBUG storage_buffer_test odoo.sql_db: SQL LOG from:
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_quant_package: 0:00:00.000675/1
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_package_storage_type: 0:00:00.000445/2
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: product_packaging: 0:00:00.000506/2
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_picking: 0:00:00.000674/2
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: res_users: 0:00:00.001414/2
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: res_partner: 0:00:00.001668/2
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: res_lang: 0:00:00.000850/3
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_inventory_line: 0:00:00.001044/3
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_warehouse: 0:00:00.001534/3
2021-01-22 14:36:18,183 31973 DEBUG storage_buffer_test odoo.sql_db: table: product_category: 0:00:00.001006/4
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_location_route: 0:00:00.001318/4
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: ir_config_parameter: 0:00:00.001382/4
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_move: 0:00:00.001505/4
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_location_storage_type: 0:00:00.000819/5
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: product_template: 0:00:00.001904/5
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_storage_location_sequence: 0:00:00.002430/5
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: ir_sequence: 0:00:00.001826/6
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: product_product: 0:00:00.002585/7
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_move_line: 0:00:00.003467/8
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: uom_uom: 0:00:00.001723/9
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: ir_model_fields: 0:00:00.002355/9
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_rule: 0:00:00.002564/9
2021-01-22 14:36:18,184 31973 DEBUG storage_buffer_test odoo.sql_db: table: decimal_precision: 0:00:00.001615/10
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_quant: 0:00:00.002618/10
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_picking_type: 0:00:00.002570/11
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: ir_property: 0:00:00.005719/12
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_location: 0:00:00.015621/60
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: ir_model_data: 0:00:00.014757/64
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: SUM from:0:00:00.076594/382 [394]
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: SQL LOG into:
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_location_allowed_location_storage_type_rel: 0:00:00.000363/1
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_location: 0:00:00.001017/1
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_rule: 0:00:00.001529/1
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: product_product: 0:00:00.001014/2
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: product_packaging: 0:00:00.001367/2
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_production_lot: 0:00:00.001439/2
2021-01-22 14:36:18,185 31973 DEBUG storage_buffer_test odoo.sql_db: table: ir_property: 0:00:00.001481/2
2021-01-22 14:36:18,186 31973 DEBUG storage_buffer_test odoo.sql_db: table: product_template: 0:00:00.001832/2
2021-01-22 14:36:18,186 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_move_line: 0:00:00.002347/2
2021-01-22 14:36:18,186 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_quant: 0:00:00.003306/2
2021-01-22 14:36:18,186 31973 DEBUG storage_buffer_test odoo.sql_db: table: stock_move: 0:00:00.004143/3
2021-01-22 14:36:18,186 31973 DEBUG storage_buffer_test odoo.sql_db: SUM into:0:00:00.019836/382 [394]
```

So 28 queries less when we do not store them.

But also:

```
2021-01-22 15:14:37,871 1 ERROR foo odoo.osv.expression: Non-stored field stock.location.location_will_contain_product_ids cannot be searched.
2021-01-22 15:14:37,871 1 ERROR foo odoo.osv.expression: Non-stored field stock.location.location_will_contain_product_ids cannot be searched.
2021-01-22 15:14:41,092 1 ERROR foo odoo.osv.expression: Non-stored field stock.location.location_will_contain_product_ids cannot be searched.
2021-01-22 15:14:41,093 1 ERROR foo odoo.osv.expression: Non-stored field stock.location.location_will_contain_product_ids cannot be searched.
```

IIRC the cache was used to make the computation of the domain for the location selection faster in the destination location, as we have to recompute the value for all the allowed locations.

![Image of the domain to select a destination location](https://user-images.githubusercontent.com/417223/105506484-64400b00-5cca-11eb-98f5-56d700923ff6.png)

guewen commented:
Also, if you don't store it, you can't search for candidates, you'd have to compute all of them for all locations (we can run a single SQL query to get all the allowed locations at once).
With the stored computed field, you re-compute only those that change, if it was not stored, you'd have to compute this for *all* sublocations to filter out the allowed locations. Storing these fields is correct to me.

@jbaudoux

lmignon reviewed (approved):
LGTM (Code review only). I think that one benefit introduced by this PR is that these 2 fields will no more be computed on heavily used locations like the Customer location...

guewen commented:
>  I think that one benefit introduced by this PR is that these 2 fields will no more be computed on heavily used locations like the Customer location...

You are right, and then I can even improve further by explicitly ignoring locations != "internal", which will avoid checking if we have location storage types on the location.

Commit:
Exclude non-internal locations for storage types computed fields

We'll never do put-away in special locations (supplier, customer, ...),
only in internal locations. We can then ignore computation for the "will
contain product/lot" fields and for the is empty field for special
locations.

guewen commented:
@jbaudoux @simahawk can you pass by for an update on your reviews?

simahawk commented:
/ocabot merge minor
